### PR TITLE
Open-loop odometry

### PR DESCRIPTION
**Pull request by efernandez**
The purpose of this PR is to support open loop odometry when there are no encoders.
It basically adds:
- open_loop param
- extends `Odometry` class to update using the last linear and angular velocities received by the controller, i.e. set the odometry twist directly with then and integrate the position (x, y)

### CODE REVIEW

File: diff_drive_controller/src/diff_drive_controller.cpp [hunk 1]
```diff
@@ -108,7 +108,8 @@ static bool getWheelRadius(const boost::shared_ptr<const urdf::Link>& wheel_link
 namespace diff_drive_controller{
 
   DiffDriveController::DiffDriveController()
-    : command_struct_()
+    : open_loop_(false)
```
Comment (po1) [inline diff_drive_controller/src/diff_drive_controller.cpp:111]: It is better style to add new parameters at the bottom. This _can_ help prevent ABI breaks.

Comment (efernandez) [inline diff_drive_controller/src/diff_drive_controller.cpp:111]: I tried to put it together with other odometry related attributes in the header file.
It just happened to be before `command_struct_`, so I had to put it at the beginning.

File: diff_drive_controller/src/odometry.cpp [hunk 2]
```diff
@@ -95,6 +95,21 @@ namespace diff_drive_controller
     return true;
   }
 
+  bool Odometry::update_open_loop(double linear, double angular, const ros::Time& time)
```
Comment (po1) [inline diff_drive_controller/src/odometry.cpp]: I don't remember why, but Hilario might be able to tell you whether you should use `time` or directly `period`.

@hilario

File: diff_drive_controller/src/diff_drive_controller.cpp (hunk 1 above)
Comment (po1) [inline diff_drive_controller/src/diff_drive_controller.cpp:111]: no problem. This is not really a library anyway, so nothing builds against this API.

**Comment by efernandez** [comment]
Nice refactoring @bmagyar 
:+1: 

Now we use `period` instead of `time` for the odometry integration.
@po1 What do you think?
I add @adolfo-rt to the loop.

File: diff_drive_controller/include/diff_drive_controller/odometry.h [hunk 3]
```diff
@@ -76,7 +78,16 @@ namespace diff_drive_controller
...
+    void update(double left_pos, double right_pos, const ros::Duration &period);
+
+    /**
+     * \brief Updates the odometry class with latest velocity command
+     * \param linear  Linear velocity [m/s]
+     * \param angular Angular velocity [rad/s]
+     * \param time    Current time
+     * \return true if the odometry is actually updated
+     */
+    void update_open_loop(double linear, double angular, const ros::Duration &period);
```
Comment (adolfo-rt) [inline diff_drive_controller/include/diff_drive_controller/odometry.h]: updateOpenLoop

**Comment by lucamarchionni** [comment]
Period is constant in every control loop execution.
It happens that there is jitter and using time differences in between control loop executions could be more accurate.

**Comment by adolfo-rt** [comment]
Could you provide a brief summary of the PR, to better match the code with the expectations of the PR? (and to make sure that no extra changesets made it through by mistake).

Also, please mention API additions and changes. This PR changes public interfaces, but the affected classes might be internal.

File: diff_drive_controller/include/diff_drive_controller/odometry.h (hunk 3 above)
Comment (po1) [inline diff_drive_controller/include/diff_drive_controller/odometry.h]: +1
also, same as above for `update()`

File: diff_drive_controller/include/diff_drive_controller/odometry.h [hunk 4]
```diff
@@ -153,14 +173,15 @@ namespace diff_drive_controller
...
 
-    /// Current timestamp:
-    ros::Time timestamp_;
-
     /// Current pose:
     double x_;        //   [m]
     double y_;        //   [m]
     double heading_;  // [rad]
 
+    /// Current velocity (for open loop mode):
```
Comment (po1) [inline diff_drive_controller/include/diff_drive_controller/odometry.h]: after Bence's refactoring it's now for both open loop and closed loop

File: diff_drive_controller/include/diff_drive_controller/odometry.h [hunk 5]
```diff
@@ -76,7 +78,16 @@ namespace diff_drive_controller
      * \param time      Current time
      * \return true if the odometry is actually updated
      */
-    bool update(double left_pos, double right_pos, const ros::Time &time);
+    void update(double left_pos, double right_pos, const ros::Duration &period);
```
Comment (po1) [inline diff_drive_controller/include/diff_drive_controller/odometry.h]: [pedantic] in the implementation, the `&` is sticked to the type, not to the parameter.

**Comment by po1** [comment]
@bmagyar thanks for the refactoring, but the title of the commit is misleading, since you also changed the behavior (time -> period)

**Comment by efernandez** [comment]
@adolfo-rt  The purpose of this PR was originally to support open loop odometry when there are no encoders.
It basically adds:
- `open_loop` param
- extends `Odometry` class to update using the last linear and angular velocities received by the controller, i.e. set the odometry twist directly with then and integrate the position `(x, y)`

Some refactoring was done later that basically simplifies setting the odometry message and tf with the odometry values (especially the `getLinear` and `getAngular` methods).

We also changed the behaviour to use `period` instead of `time`, but that might go to another PR IMHO.
@lucamarchionni Here we're basically asking what should we use. `period` to filtered out the jitter (as it's constant), or time`?

**Comment by efernandez** [comment]
@adolfo-rt This PR only affects internal classes. It adds `updateOpenLoop` method and a few attributes to the class.

**Comment by po1** [comment]
It also adds a rosparam, that is external API that has to be documented.
If this is to be merged upstream, and it should be, then a unit test exercising the new feature would be welcome.

**Comment by efernandez** [comment]
@po1 I think the simple test I've added should be enough.
Could you give it a try?

**Comment by efernandez** [comment]
Testing on REEM-H2.
With `open_loop` it's easy to see that the odometry has no noise, while in close loop it has. :)
This is normal and expected, and it's a signal that the changes are OK.
Anyway, I still have to test a little more, moving the robot, and also using close loop, to verify that there's no regression.

**Comment by efernandez** [comment]
It works, both open and close loops.
@lucamarchionni @po1 @bmagyar @adolfo-rt Any other comment or something missed before merging?

**Comment by po1** [comment]
all good for me

**Comment by efernandez** [comment]
After talking with @lucamarchionni we have decided that it's better to use the `time` difference instead of the `period`.
I'll do and test that.

**Comment by po1** [comment]
@efernandez I'd be curious to have a short summary of the motivation for reverting to `time`

**Comment by efernandez** [comment]
I've gone back to `time` for the odometry computations.

I've also detected a potential bug with the `timestamp_` not initialized correctly.
Therefore, I've added an `init` method in a separate commit.

@po1 @bmagyar Could you give it a look?
I'll test it on the robot.

**Comment by efernandez** [comment]
BTW, all (8/8) unit tests passed.

**Comment by bmagyar** [comment]
The reason for using `time` is that `period` is fixed, the theoretical update cycle of the controller, while the update function might be called with slight differences. This could potentially introduce small jitter in the controller, while another reason for not changing could be [IfItAintBrokeDontFixIt](http://c2.com/cgi/wiki?IfItAintBrokeDontFixIt).

The init function definitely makes things better, now we won't have a big bump upon the first update. This wasn't causing problems on REEM but might do on other platforms, better be safe.

**Comment by efernandez** [comment]
Last changes tested on REEM-H2 robot.
Everything looks OK.

@po1 I've followed @lucamarchionni recommendation to use the `time` because it's more accurate, especially if the real time system doesn't really follow the `period`.

@bmagyar @adolfo-rt If you find it right now, I'll merge it.

**Comment by bmagyar** [comment]
I'd squash the 2 commits about cosmetics, and rebase the rest, but the contents are fine.

**Comment by bmagyar** [comment]
+1

**Comment by efernandez** [comment]
Cosmetic changes squashed into one.

I don't feel like merging my own PR...

**Comment by po1** [comment]
@bmagyar  @efernandez IIRC that was exactly the motivation to use `period` in the first place. The jitter introduced by the controller in the odometry will be amplified in the computation of the velocities, which would be way smoother if you used the a priori knowledge about the frequency of the controller, which is `period`. I understand that it works that way for now, but I seem to remember that it works the other way around in our walking controller, which is far more sensible to jitter...

@efernandez I suggest you give a try to both options, especially on a system that has a very low resolution odometry, if you see what I refer to.

In the mean time, let's merge this and open more PRs if needed.

**Comment by po1** [comment]
@bmagyar in the future, please don't label a commit as 'cosmetic' if it contains changes in the API (even internal) and the logic.

**Pull request merged**